PathColumn : Resize modes

murraystevenson opened the pull request:
Originating from attempts to stretch the width of the HierarchyView's Name column after the addition of the Inclusions and Exclusions columns, this PR adds flags to PathColumn allowing a resize mode to be defined, which is then converted to Qt header resize behaviour when the column is added to a PathListingWidget.

The PathListingWidget's `defaultNameColumn` now defaults to ResizeStretch, allowing it to make use of available space, rather than the previous behaviour of whichever column was added last to a PathListingWidget stretching via Qt's `stretchLastSection` property. `stretchLastSection` is now disabled when there are columns in the PathListingWidget specifically configured to stretch, but enabled for PathListingWidgets containing none, to preserve the existing behaviour. There's also a small optimization to `updateColumnWidths()` as we should be able to ignore columns that aren't interactively resizable.

To demonstrate wider use, additional commits have been included to set ResizeStretch on the ShaderUI, _HistoryWindow and LightEditor's name column equivalents. Of these, the LightEditor may benefit from a custom resize behaviour where the column is permitted to stretch while also maintaining a minimum width, as the potentially large and varying number of columns in the LightEditor can cause the `LocationNameColumn` to shrink down to nothing when there isn't enough available space to display all columns. The Catalogue's Name column could suffer a similar fate. @ericmehl, maybe you could take this for a spin and see what you think?

John, please let me know any thoughts on the structure here. I went with the generic `flags` in case we need to support other behaviours/interesting quirks of Qt in the future, but maybe it's preferable to have a more specific `ResizeMode` enum and deal with any future cases if/when they arise. I may also need a hand with the boost::python incantation required to get the enum binding first for use in PathColumn's `init`...

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -902,20 +909,22 @@ def updateColumnWidths( self ) :
 			self.__columnWidthAdjustments = [ 0 ] * numColumnsToResize
 
-		del self.__idealColumnWidths[:]
+		self.__idealColumnWidths = [ 0 ] * numColumnsToResize
 		for i in range( 0, numColumnsToResize ) :
+			if header.sectionResizeMode( i ) == QtWidgets.QHeaderView.ResizeMode.Interactive :
+				idealWidth = max( header.sectionSizeHint( i ), self.sizeHintForColumn( i ) )
+				self.__idealColumnWidths[i] = idealWidth
```
It took me a while to figure out that it should be OK to allow values of `0` in `__idealColumnWidths` because `__sectionResized` shouldn't be called for non-interactive-resize sections. It still feels a bit fragile though. The existing behaviour where we clear all the `__columnWidthAdjustments` when columns are changed also feels a bit weird - if we just added a column on the end, should that really invalidate the adjustments? It feels a bit like what we want is a dictionary mapping from column to adjustment, but that could be a bit fiddly.

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -902,20 +909,22 @@ def updateColumnWidths( self ) :
-			self.__idealColumnWidths.append( idealWidth )
-
-			header.resizeSection( i, idealWidth + self.__columnWidthAdjustments[i] )
+				header.resizeSection( i, idealWidth + self.__columnWidthAdjustments[i] )
```
How come we only call `resizeSection` for interactive sections now? I guess it's because Qt is automatically resizing to contents as necessary for the other modes?

johnhaddon commented on src/GafferSceneUIModule/HierarchyViewBinding.cpp:
```diff
@@ -510,8 +510,8 @@ class InclusionsColumn : public PathColumn
 
-		InclusionsColumn( ContextPtr context )
-			:	PathColumn(), m_context( context )
+		InclusionsColumn( ContextPtr context, unsigned flags = PathColumn::ResizeToContents )
```
Is there any real benefit to exposing this as a constructor argument? Perhaps InclusionsColumn knows best and should just hardcode itself to ResizeToContents?

johnhaddon commented on include/GafferUI/PathColumn.h:
```diff
@@ -63,6 +63,26 @@ class GAFFERUI_API PathColumn : public IECore::RefCounted, public Gaffer::Signal
+		{
+			ResizeInteractive = 0,
+			ResizeStretch = 1,
+			ResizeFixed = 2,
```
Does `ResizeFixed` make sense in an API where we don't provide any methods to allow that fixed width to be specified?

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -61,7 +61,7 @@ class PathListingWidget( GafferUI.Widget ) :
 
 	## A collection of handy column definitions for FileSystemPaths
-	defaultNameColumn = StandardColumn( "Name", "name" )
+	defaultNameColumn = StandardColumn( "Name", "name", flags = GafferUI.PathColumn.Flags.ResizeStretch )
```
I'm not 100% sure about this - here are two downsides from the File Open dialogue that we might want to consider :

- If you have a name that is too wide to see, you can no longer resize the Name column to see it.
- Resizing the Owner column feels kindof weird - you drag the right edge to the right and it stays put while the left edge moves to the left.

johnhaddon commented on src/GafferSceneUIModule/LightEditorBinding.cpp:
```diff
@@ -84,8 +84,8 @@ class LocationNameColumn : public StandardPathColumn
 
-		LocationNameColumn()
-			:	StandardPathColumn( "Name", "name" )
+		LocationNameColumn( unsigned flags = PathColumn::ResizeStretch )
```
This might just be because I've recently inherited a very wide monitor, but I'm not sure this is the ideal setting - it can introduce a lot of dead space between the light names and the parameters :

![image](https://user-images.githubusercontent.com/1133871/210400743-5bc589ee-4dc0-4499-a3e3-253e047b64fd.png)

johnhaddon reviewed (commented):
Thanks Murray! The HierarchyView is definitely much improved by this, but I'm not as convinced by the changes to the other PathListingWidgets, or by the lifting of the Qt ResizeMode one-to-one into the Gaffer API. Some of this might just be that I'm used to the existing quirks and just need to adjust to the change, but I think it's worth taking this moment to step back a bit and consider what we think the ideal set of behaviours might be, before going back into the Qt world to try to implement them. Stuff to consider :

- Is the old `__columnWidthAdjustments` stuff worth it still?
- Might it be desirable that a column could be both stretchy _and_ interactively resizeable? (for the File Open case, I think I would say yes).
- What is the minimum Qt-agnostic API we might want if we were designing from scratch?

Cheers...
John

murraystevenson commented:
Here’s an alternate take on the column resize behaviour that foregoes the various Qt modes and implements stretchy behaviour on top of Qt’s regular Interactive column. This leaves us with two resize modes, `Interactive` for the existing user adjustable column behaviour, and `Stretch` for user adjustable but automatically stretching to fill available space. `Stretch` attempts to address the caveats of Qt’s native stretch behaviour - the lack of interactivity and the odd “resizing from the wrong side” behaviour of columns to the right of a stretchable column. Thoughts on the general behaviour are very welcome, we may still want to add a fixed width mode in the future though I don't think we have a current use-case for it...

This also adds the suggested dictionary mapping of `__idealColumnWidths` and `__columnWidthAdjustments`, which is a definite usability improvement over the existing approach.

I’ve dropped the commits stretching the Name column equivalents in the LightEditor, HistoryWindow, and ShaderUI. Those don’t benefit much from the stretch behaviour - especially the LightEditor on certain spaceship sized monitors… 🚀  For testing purposes I’ve replaced those with the LocalDispatchUI's Name column, and made the Catalogue's Description column stretch in order to test multiple stretchy columns in a single PathListingWidget - both Name and Description should stretch there.

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -1030,5 +1025,14 @@ def __sectionResized( self, index, oldWidth, newWidth ) :
-		if len( self.__idealColumnWidths ) > index :
-			self.__columnWidthAdjustments[index] = newWidth - self.__idealColumnWidths[index]
+		column = self.__getColumn( index )
+		self.__columnWidthAdjustments[column] = newWidth - self.__idealColumnWidths[column]
```
This is what fixes the annoying "jump back to default width when you change tabs" thing in the LightEditor, right? Might be worth a mention in Changes.md.

murraystevenson replied:
Added in 8d39641.

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -97,7 +98,7 @@ def __init__(
 		displayMode = DisplayMode.List,
 		sortable = True,
-		horizontalScrollMode = GafferUI.ScrollMode.Never,
+		horizontalScrollMode = GafferUI.ScrollMode.Automatic,
```
Technically this is an API breaking change that we should mention in Changes.md. Either that, or we change it back and then change the places where we're constructing a PathListingWidget. I think my general aversion to horizontal scrollbars clouds my judgement here, so I'll leave it to you...

murraystevenson replied:
Makes sense. `Automatic` was useful for testing but I've reverted back to `Never` in 64482f7. There might be value in enabling `Automatic` on the HierarchyView once we start adding more columns, but I think we can leave this be for the moment...

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -1028,6 +1081,108 @@ def __sectionResized( self, index, oldWidth, newWidth ) :
+	def __resizeLastColumnToAvailableWidth( self, minimumSuggestedWidth = 0 ) :
+
+		header = self.header()
+		if header.stretchLastSection() :
+			# Qt's behaviour will override and extend the last section for us automatically
+			return
```
Is this redundant, since we call `setStretchLastSection()` in `_TreeView.__init__`?

murraystevenson replied:
Indeed, redundant now we're forcing that off in `__init__`. Removed in 63a7d55.

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -903,14 +936,34 @@ def updateColumnWidths( self ) :
 		header = self.header()
-		numColumnsToResize = header.count() - 1 # leave the last section alone, as it's expandable
+		numColumnsToResize = header.count()
+		if header.stretchLastSection() or self.getStretchLastColumn() :
```
I think we know that `stretchLastSection()` is `False` already don't we, since we hardcoded it in `_TreeView.__init__`?

murraystevenson replied:
Removed in 63a7d55.

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -302,6 +303,8 @@ def setColumns( self, columns ) :
 			return
 
 		_GafferUI._pathListingWidgetSetColumns( GafferUI._qtAddress( self._qtWidget() ), columns )
+		self._qtWidget().setStretchLastColumn( not any( c.getResizeMode() == GafferUI.PathColumn.ResizeMode.Stretch for c in columns ) )
```
It took me a minute to realise that we weren't controlling a Qt behaviour here. Since the value is always derived from the column modes, perhaps things are actually clearer if we omit the `[sg]etStretchLastColumn` methods and just let `_TreeView` determine what to do based on the columns directly? I think we're basically duplicating state at the moment, which makes it a bit harder to follow the logic. If we remove this, then I think that means `_TreeView` gets to encapsulate all the resize logic, rather than sharing it with PathListingWidget.

murraystevenson replied:
Good point! I took a stab at this in a24f839.

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -903,14 +936,34 @@ def updateColumnWidths( self ) :
+				idealWidth = max( header.sectionSizeHint( i ), self.sizeHintForColumn( i ) )
+				self.__idealColumnWidths[column] = idealWidth
+				adjustedWidth = idealWidth + self.__columnWidthAdjustments[column]
+				if( i == header.count() - 1 ) :
```
Do we actually need the `()` here, or is C++ just leaking into Python?

murraystevenson replied:
:upside_down_face:  Addressed in c43b367.

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -1028,6 +1081,108 @@ def __sectionResized( self, index, oldWidth, newWidth ) :
+	def __resizeStretchColumns( self ) :
+
+		header = self.header()
+		if header.stretchLastSection() :
+			# Qt's behaviour will override and result in the last section taking
+			# all available space so there's no need to calculate our own widths.
+			return
```
Another one of those ones that I think should always be `False` - maybe I missed something?

murraystevenson replied:
Removed in 63a7d55.

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -1036,3 +1191,16 @@ def __getColumn( self, index ) :
 
 		return columns[index]
+
+	def __getColumnResizeMode( self, index ) :
+
+		columns = _GafferUI._pathListingWidgetGetColumns( GafferUI._qtAddress( self ) )
+
+		if index >= len( columns ) :
+			raise IndexError( index )
+
+		if index == len( columns ) - 1 and self.getStretchLastColumn() :
+			# override the last column to stretch
+			return GafferUI.PathColumn.ResizeMode.Stretch
+
+		return columns[index].getResizeMode()
```
Probably not important, but `__getColumn()` and `__getColumnResizeMode()` lead towards accidentaly `O( n^2 )` because they both retrieve all `N` columns and will likely be called `N` times. Probably the number of columns is small enough that it doesn't matter, but I do wonder what some of the code might look like if it retrieved the columns once and then looped over them.

murraystevenson replied:
Yeah, given the relatively low number of columns we're anticipating it's probably not the end of the world, but I took another look at this and reworked things in 419c73.

johnhaddon reviewed (commented):
Thanks for taking this one on Murray - it's always scary getting up this close to Qt!

Playing with it, I noticed improvements in all the major PathListingWidget UIs - LightEditor, HierarchyView, Catalogue and PathChooserDialogue, which is great. I do wonder a little about how easy it is to get into an unwanted-horizontal-scrollbar situation now, but that might just be my personal preferences. There are so many interconnected parts here, and you've documented the current strategy well, so I don't think there's any value in trying to change tack at this point. Best to get it onto `main` so we can all get some practical experience with it.

My inline comments are all pretty minor, and might not all be relevant either, so hopefully we can get this across the line quickly...

murraystevenson commented on include/GafferUI/PathColumn.h:
```diff
@@ -63,6 +63,24 @@ class GAFFERUI_API PathColumn : public IECore::RefCounted, public Gaffer::Signal
 		IE_CORE_DECLAREMEMBERPTR( PathColumn )
 
+		/// Defines the UI size behaviour of the column.
+		enum ResizeMode
```
Consistency sounds good to me! I've gone with `SizeMode` in 86266c4, f0e451f, 0f751cb & 44e92fc.

murraystevenson commented:
Feedback addressed and rebased to fix the Changes.md conflict. There's some CI weirdness to look into in the morning and I'll update the references to `resizeMode` in the commit messages once everything's squashed down.

johnhaddon reviewed (approved):
LGTM. CI weirdness was unrelated to this PR.
